samples/rust/rust_debugfs.rs

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Rust misc device that reports debug information with a seq_file entry in
+//! debugfs.
+
+use kernel::prelude::*;
+use kernel::{
+    debugfs::{debugfs_create, DebugFsDirectory},
+    file::{self, File},
+    io_buffer::IoBufferWriter,
+    miscdev,
+    str::CString,
+    sync::{Mutex, Ref, RefBorrow, UniqueRef},
+};
+
+module! {
+    type: RustMiscdev,
+    name: b"rust_debugfs",
+    author: b"Rust for Linux Contributors",
+    description: b"Sample Rust miscellaneous device with a debugfs entry",
+    license: b"GPL v2",
+}
+
+#[derive(Clone, Copy)]
+struct SharedStateInner {
+    read_count: usize,
+}
+
+struct SharedState {
+    inner: Mutex<SharedStateInner>,
+}
+
+impl SharedState {
+    fn try_new() -> Result<Ref<Self>> {
+        let mut state = Pin::from(UniqueRef::try_new(Self {
+            // SAFETY: `mutex_init!` is called below.
+            inner: unsafe { Mutex::new(SharedStateInner { read_count: 0 }) },
+        })?);
+
+        // SAFETY: `inner` is pinned when `state` is.
+        let pinned = unsafe { state.as_mut().map_unchecked_mut(|s| &mut s.inner) };
+        kernel::mutex_init!(pinned, "SharedState::inner");
+
+        Ok(state.into())
+    }
+}
+
+struct Token;
+#[vtable]
+impl file::Operations for Token {
+    type Data = Ref<SharedState>;
+    type OpenData = Ref<SharedState>;
+
+    fn open(shared: &Ref<SharedState>, _file: &File) -> Result<Self::Data> {
+        Ok(shared.clone())
+    }
+
+    fn read(
+        shared: RefBorrow<'_, SharedState>,
+        _: &File,
+        _data: &mut impl IoBufferWriter,
+        _offset: u64,
+    ) -> Result<usize> {
+        let mut inner = shared.inner.lock();
+        inner.read_count += 1;
+        Ok(0)
+    }
+}
+
+struct DebugFsToken;
+#[vtable]
+impl file::Operations for DebugFsToken {
+    type OpenData = Ref<SharedState>;
+    type Data = Ref<SharedState>;
+
+    fn open(shared: &Ref<SharedState>, _file: &File) -> Result<Self::Data> {
+        Ok(shared.clone())
+    }
+
+    fn read(
+        shared: RefBorrow<'_, SharedState>,
+        _: &File,
+        data: &mut impl IoBufferWriter,
+        offset: u64,
+    ) -> Result<usize> {
+        if offset != 0 {
+            return Ok(0);
+        }
+
+        let read_count = shared.inner.lock().read_count;
+        let string = CString::try_from_fmt(fmt!("Debugfs file read count: {}\n", read_count))?;
+        data.write_slice(string.as_bytes())?;
```

Comment:
Note that here we're first creating the `CString` and then copying it into the `IoBufferWriter` because I don't think we currently have a way to directly write format strings to an `IoBufferWriter` (unless I'm missing something?).